File: problems/Remove All Adjacent Duplicates in String II/removeDuplicates.cs
public class Solution {
    public string RemoveDuplicates(string s, int k) {
        var stack = new Stack<(char, int)>();

        foreach (char letter in s) {
            if (0 < stack.Count && letter == stack.Peek().Item1) {
                if (k == 1 + stack.Peek().Item2) {
                    while (0 < stack.Count && letter == stack.Peek().Item1) {
                        stack.Pop();
                    }
                } else {
                    stack.Push((letter, 1 + stack.Peek().Item2));
                }

            } else {
                stack.Push((letter, 1));
            }
        }

        var sb = new StringBuilder();

        foreach (var letter in stack) {
            sb.Insert(0, letter.Item1);
        }

        return sb.ToString();
    }
}

File: problems/Remove All Adjacent Duplicates in String II/removeDuplicates.java
class Solution {
    public String removeDuplicates(String s, int k) {
        StringBuilder sb = new StringBuilder(s);
        Stack<Integer> store = new Stack<>();

        for (int idx = 0; sb.length() > idx; ++idx) {
            if (0 == idx || sb.charAt(idx - 1) != sb.charAt(idx)) {
                store.push(1);
            } else {
                int count = 1 + store.pop();

                if (k == count) {
                    sb.delete(1 + idx - k, 1 + idx);
                    idx = idx - k;
                } else {
                    store.push(count);
                }
            }
        }

        return sb.toString();
    }
}
